pilot/pkg/xds/ecds.go

```diff
@@ -71,5 +114,87 @@ func (e *EcdsGenerator) Generate(proxy *model.Proxy, push *model.PushContext, w
 			Resource: util.MessageToAny(c),
 		})
 	}
+
 	return resources, model.DefaultXdsLogDetails, nil
 }
+
+func (e *EcdsGenerator) GeneratePullSecrets(proxy *model.Proxy, updatedSecrets map[model.ConfigKey]struct{}, secretResources []SecretResource,
+	secretController credscontroller.Controller, req *model.PushRequest) map[string][]byte {
+	if proxy.VerifiedIdentity == nil {
+		log.Warnf("proxy %s is not authorized to receive secret. Ensure you are connecting over TLS port and are authenticated.", proxy.ID)
+		return nil
+	}
+
+	results := make(map[string][]byte)
+	for _, sr := range secretResources {
+		if updatedSecrets != nil {
+			if _, found := updatedSecrets[model.ConfigKey{Kind: gvk.Secret, Name: sr.Name, Namespace: sr.Namespace}]; !found {
+				// This is an incremental update, filter out credscontroller that are not updated.
+				continue
+			}
+		}
+
+		cred, err := secretController.GetDockerCredential(sr.Name, sr.Namespace)
```

Comment:
does it mean that the proxy can get secret from any namespaces?

Reply:
this is hack, `sr.Namespace` actually come form `plugin.Namespace`
```
wasmPlugin.ImagePullSecret = toSecretResourceName(wasmPlugin.ImagePullSecret, plugin.Namespace)
```

Reply:
if `wasmplugin` config in `rootNamespace`, all sidecar will pull secret from that namespace, there's a risk of secret leak.

Reply:
This is something similar to sidecar sds, if a user can define DR for a sidecar, and the sidecar have permission to read the cert defined.

Reply:
Yeah, this is a known behavior. We need to make it clear at the reference doc that referencing a secret at WasmPlugin == secret will be distributed to the proxies that it applies to.

Reply:
IMO, we should not support mesh global wasm plugin in root namespace.  In the current case, it could not get image pull secret as the `namespace ` may differ.

Reply:
@howardjohn you opinion?

Reply:
Sorry just noticed this comment. Then how do you achieve global WasmPlugin config? Seems really hard to manage if secret needs to be copied into every namespace?

Reply:
Alternative is to let control plane pull and push Wasm bytes, which could take some effort to implement.

If this is a concern, we probably could have an env var at pilot to disable secret distribution from root namespace and only allow it when user explicitly set it.